### PR TITLE
Implement node Allocatable

### PR DESCRIPTION
**Pull request by timstclair**
Add `kube-reserved` and `system-reserved` flags for configuration
reserved resources for usage outside of kubernetes pods. Allocatable is
provided by the Kubelet according to the formula:

```
Allocatable = Capacity - KubeReserved - SystemReserved
```

Issue: https://github.com/kubernetes/kubernetes/issues/13984

Proposal: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/node-allocatable.md

Includes commits from dependent PRs: https://github.com/kubernetes/kubernetes/pull/18189 and https://github.com/kubernetes/kubernetes/pull/18557

Remaining work: set default value of kube-reserved based on predicted usage

/cc @kubernetes/goog-node

### CODE REVIEW

**Comment by dchen1107** [comment]
Issue: #13984

**Comment by timstclair** [comment]
> Shouldn't default is no reservation for this one too, at least for now?

Yeah, makes sense. I pulled out the defaulting code.

**Comment by derekwaynecarr** [comment]
I would like time to do an initial review of this (eta 12/15)

**Comment by timstclair** [comment]
BTW, all this PR does is add 2 flags (`--system-reserved` and `--kube-reserved`), and fill in `Allocatable` in the node status by subtracting the values of those flags from `Capacity`. It doesn't try to do any enforcement.

**Comment by derekwaynecarr** [comment]
@timstclair - and per our slack chat for those that did not see it, cgroup enforcement is not in scope for 1.2 for `system-reserved` and `kubernetes-reserved`, but only local node accounting to see actual usage is intended.

**Comment by timstclair** [comment]
Not quite: cgroup enforcement of `kubernetes-reserved` may be in scope for 1.2, but needs more discussion.

**Comment by derekwaynecarr** [comment]
@timstclair : fair enough, but if that is an explicit requirement, we need to take into account
https://github.com/kubernetes/kubernetes/pull/17688

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by dchen1107** [comment]
LGTM

**Comment by timstclair** [comment]
Thanks. This is still blocked by https://github.com/kubernetes/kubernetes/pull/18557 though.

File: docs/admin/kubelet.md [hunk 1]
```diff
@@ -99,6 +99,7 @@ kubelet
       --image-gc-low-threshold=80: The percent of disk usage before which image garbage collection is never run. Lowest disk usage to garbage collect to. Default: 80%
       --kube-api-burst=10: Burst to use while talking with kubernetes apiserver
       --kube-api-qps=5: QPS to use while talking with kubernetes apiserver
+      --kube-reserved=: A set of ResourceName=ResourceQuantity pairs that describe resources reserved for kubernetes system components. Currently only cpu and memory are supported. [default=none]
```
Comment (vishh) [inline docs/admin/kubelet.md]: Should we give an example on how to use this flag?

File: pkg/api/types.go [hunk 2]
```diff
@@ -1469,8 +1469,10 @@ type NodeSystemInfo struct {
 
 // NodeStatus is information about the current status of a node.
 type NodeStatus struct {
-	// Capacity represents the available resources of a node.
+	// Capacity represents the total resources of a node.
 	Capacity ResourceList `json:"capacity,omitempty"`
+	// Allocatable represents the resources of a node that are available for scheduling.
+	Allocatable ResourceList `json:"allocatable,omitempty"`
```
Comment (vishh) [inline pkg/api/types.go]: Should we clarify that `Capacity` and `Allocatable` are not expected to change over time?

File: pkg/api/v1/types_swagger_doc_generated.go [hunk 3]
```diff
@@ -702,7 +702,8 @@ func (NodeSpec) SwaggerDoc() map[string]string {
 
 var map_NodeStatus = map[string]string{
 	"":                "NodeStatus is information about the current status of a node.",
-	"capacity":        "Capacity represents the available resources of a node. More info: http://releases.k8s.io/HEAD/docs/user-guide/persistent-volumes.md#capacity for more details.",
+	"capacity":        "Capacity represents the total resources of a node. More info: http://releases.k8s.io/HEAD/docs/user-guide/persistent-volumes.md#capacity for more details.",
```
Comment (vishh) [inline pkg/api/v1/types_swagger_doc_generated.go]: Why are we linking to `persistent-volumes` doc?

File: cmd/kubelet/app/server.go [hunk 4]
```diff
@@ -345,6 +350,8 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.IntVar(&s.KubeAPIBurst, "kube-api-burst", s.KubeAPIBurst, "Burst to use while talking with kubernetes apiserver")
 	fs.BoolVar(&s.SerializeImagePulls, "serialize-image-pulls", s.SerializeImagePulls, "Pull images one at a time. We recommend *not* changing the default value on nodes that run docker daemon with version < 1.9 or an Aufs storage backend. Issue #10959 has more details. [default=true]")
 	fs.BoolVar(&s.ExperimentalFlannelOverlay, "experimental-flannel-overlay", s.ExperimentalFlannelOverlay, "Experimental support for starting the kubelet with the default overlay network (flannel). Assumes flanneld is already running in client mode. [default=false]")
+	fs.Var(&s.SystemReserved, "system-reserved", "A set of ResourceName=ResourceQuantity pairs that describe resources reserved for non-kubernetes components. Currently only cpu and memory are supported. [default=none]")
```
Comment (vishh) [inline cmd/kubelet/app/server.go]: What values and units are acceptable?

**Comment by vishh** [comment]
Added a few nits. Feel free to resolve them in a separate PR if that's easier.

File: docs/admin/kubelet.md (hunk 1 above)
Comment (timstclair) [inline docs/admin/kubelet.md]: Do you mean an example of the syntax `(e.g. cpu=200m,memory=150G)`?

File: pkg/api/types.go (hunk 2 above)
Comment (timstclair) [inline pkg/api/types.go]: I'd prefer we don't make that assumption, unless there's a specific requirement. I suppose we could make the weak assertion that "Changes are relatively infrequent". But I think that by having it in status (and not spec), we're suggesting that it can change.

File: pkg/api/v1/types_swagger_doc_generated.go (hunk 3 above)
Comment (timstclair) [inline pkg/api/v1/types_swagger_doc_generated.go]: Good question. It was there before, probably copied from the PersistentVolumeSpec.Capacity field. Removed.

File: cmd/kubelet/app/server.go (hunk 4 above)
Comment (timstclair) [inline cmd/kubelet/app/server.go]: Is this the right place to describe that? Would you accept a link to the API documentation for resource list instead?

File: docs/admin/kubelet.md (hunk 1 above)
Comment (vishh) [inline docs/admin/kubelet.md]: Yes

File: pkg/api/types.go (hunk 2 above)
Comment (vishh) [inline pkg/api/types.go]: As of now our scheduler is built on this assumption. We will only add more
functionality based on this assumption.
If nodes can also vertically scale, how about making that change to the API
after ensuring that all the components can handle that change?

File: cmd/kubelet/app/server.go (hunk 4 above)
Comment (vishh) [inline cmd/kubelet/app/server.go]: Yeah. Any solution that will help our users is fine by me.

File: pkg/api/types.go (hunk 2 above)
Comment (timstclair) [inline pkg/api/types.go]: Ok. I just realized this PR doesn't actually touch types.go, so I'll send a separate PR to address these comments.

File: docs/admin/kubelet.md (hunk 1 above)
Comment (timstclair) [inline docs/admin/kubelet.md]: Done.

File: pkg/api/v1/types_swagger_doc_generated.go (hunk 3 above)
Comment (timstclair) [inline pkg/api/v1/types_swagger_doc_generated.go]: Actually, will send a separate PR.

File: cmd/kubelet/app/server.go (hunk 4 above)
Comment (timstclair) [inline cmd/kubelet/app/server.go]: Done.

**Comment by timstclair** [comment]
Pushed changes. I'll address the changes to the API documentation in a separate PR.

**Comment by vishh** [comment]
LGTM

**Comment by sttts** [comment]
LGTM from the Mesos point of view.

**Comment by timstclair** [comment]
Rebased.

**Comment by timstclair** [comment]
Squashed & rebased.  @dchen1107 - did you have any other concerns, or is this good to go?

**Comment by timstclair** [comment]
Rebased.

**Comment by dchen1107** [comment]
LGTM

File: pkg/kubelet/kubelet_test.go [hunk 5]
```diff
@@ -69,6 +69,9 @@ func init() {
 
 const testKubeletHostname = "127.0.0.1"
 
+const testReservationCPU = "200m"
```
Comment (jdef) [inline pkg/kubelet/kubelet_test.go]: nit: put these 3 consts in a `const ( ... )` block

**Comment by timstclair** [comment]
@k8s-bot unit test this

**Comment by gmarek** [comment]
failure: #19254
@k8s-bot unit test this

**Comment by gmarek** [comment]
@k8s-bot unit test this

**Comment by timstclair** [comment]
@k8s-bot unit test this

**Comment by timstclair** [comment]
@k8s-bot unit test this

**Comment by timstclair** [comment]
Rebased.
